corehq/mobile_ucr_v2_update_script.py

```diff
@@ -0,0 +1,114 @@
+# This script updates the latest versions of all apps across domains that are using v1 with no manual references
+# Steps followed
+# 1. Get All domains with mobile ucr flag enabled
+# 2. Get all apps for domain with latest released versions and mobile ucr versions that are not v2
+# 3. For each app, if it contains no V1 UCR references, update the version to 2
+
+# How to run
+# Can be run in django shell. Paste the script and execute the function process()
+# File is stored in home directory of cchq user.
+
+# V1 Examples
+# https://staging.commcarehq.org/a/test-commcare-superset/apps/view/f940fcc83bae44b8a0adaf719673fd1e/form/a0f3c5b483c645e78b6f89ee0b3b3c03/source/#form/table_child_count
+
+
+import json
+import re
+import traceback
+
+from corehq.apps.app_manager.dbaccessors import (
+    get_latest_app_ids_and_versions,
+    get_apps_by_id,
+)
+from corehq.apps.app_manager.const import MOBILE_UCR_VERSION_2
+from corehq.toggles import MOBILE_UCR
+from corehq.toggles.shortcuts import find_domains_with_toggle_enabled
+from corehq.util.log import with_progress_bar
+
+
+def save_in_log(file, data):
+    print(data)
+    file.write(data + '\n')
+
+
+def save_as_ndjson(path, data):
+    with open(path, 'a') as file:
+        print(json.dumps(data, separators=(',', ':')), file=file)
+
+
+def read_ndjson_file(path):
+    with open(path, 'r') as file:
+        return set(json.loads(line) for line in file.readlines())
+
+
+def has_non_v2_form(domain, app, log_file):
+    for form in app.get_forms():
+        save_in_log(log_file, f"Processing Form: {domain}: {form.name}")
+        # The second condition should always be False if the first one is
+        # but just as a precaution we check for it
+        if V1_FIXTURE_IDENTIFIER in form.source or RE_V1_ALL_REFERENCES.search(form.source):
+            save_in_log(log_file, f"App Contains V1 Refs: {domain}: {app.name}")
+            return True
+    return False
+
+
+def update_app(domain, app, log_file):
+    save_in_log(log_file, f"Updating App: {domain}: {app.name}: {app.id}")
+    app.mobile_ucr_restore_version = MOBILE_UCR_VERSION_2
+    app.save()
+
+
+PROCESSED_DOMAINS_PATH = '/home/zandre/cchq/updated_domains.ndjson'
+LOG_FILE = '/home/zandre/cchq/update_to_v2_ucr_script.log'
+
+V1_FIXTURE_IDENTIFIER = 'src="jr://fixture/commcare:reports'
+V1_FIXTURE_PATTERN = r'<.*src="jr://fixture/commcare:reports.*>'
+V1_REFERENCES_PATTERN = r"<.*instance\('reports'\)/reports/.*>"
+RE_V1_ALL_REFERENCES = re.compile(f"{V1_FIXTURE_PATTERN}|{V1_REFERENCES_PATTERN}")
+
+
+skip_domains = set()
+
+
+def process():
+    try:
+        processed_domains = read_ndjson_file(PROCESSED_DOMAINS_PATH)
+    except FileNotFoundError:
+        processed_domains = set()
+
+    mobile_ucr_domains = find_domains_with_toggle_enabled(MOBILE_UCR)
+
+    log_file = open(LOG_FILE, 'a')
```

Comment:
I believe this would leave the connection to the file open even after the script exits abruptly, so we should this with `with` option. This way you also don't need to worry about explicitly calling close.
`with open(LOG_FILE, 'a') as log_file:`

Does mode 'a' create the file if its missing?

Reply:
Looking through the code, I can remove the need for passing the log file around to the various helper functions. Since we already know what the file path is, we can just use this directly in the helper function for adding to the log. Addressed in [27c1c25](https://github.com/dimagi/commcare-hq/pull/35158/commits/27c1c2501790b14dfbbe049e3dff4d4c68a88656).

> Does mode 'a' create the file if its missing?

It does. [Here](https://phoenixnap.com/kb/file-handling-in-python#:~:text=Append%20mode%20adds%20information%20to,not%20clear%20a%20file's%20contents.) is an article for more info on the append mode.